mods/default/functions.lua

```diff
@@ -344,56 +344,57 @@ minetest.register_abm({
 -- Convert dirt to something that fits the environment
 --
 
+default.dirt_types = {
+	"default:dirt_with_grass",
+	"default:dirt_with_dry_grass",
+	"default:dirt_with_snow"
+}
+
 minetest.register_abm({
 	label = "Grass spread",
 	nodenames = {"default:dirt"},
-	neighbors = {
-		"default:dirt_with_grass",
-		"default:dirt_with_dry_grass",
-		"default:dirt_with_snow",
-		"group:grass",
-		"group:dry_grass",
-		"default:snow",
-	},
+	neighbors = {"air"},
 	interval = 6,
 	chance = 67,
 	catch_up = false,
 	action = function(pos, node)
-		-- Most likely case, half the time it's too dark for this.
+		-- not enough light
 		local above = {x = pos.x, y = pos.y + 1, z = pos.z}
 		if (minetest.get_node_light(above) or 0) < 13 then
 			return
 		end
 
-		-- Look for likely neighbors.
-		local p2 = minetest.find_node_near(pos, 1, {"default:dirt_with_grass",
-				"default:dirt_with_dry_grass", "default:dirt_with_snow"})
-		if p2 then
-			-- But the node needs to be under air in this case.
-			local n2 = minetest.get_node(above)
-			if n2 and n2.name == "air" then
-				local n3 = minetest.get_node(p2)
-				minetest.set_node(pos, {name = n3.name})
-				return
+		-- water above grass
+		local name = minetest.get_node(above).name
+		local def = minetest.registered_nodes[name]
+
+		if name == "ignore" or not def or def.liquidtype ~= "none" then
+			return
+		end
+
+		local curr_max, curr_type, num  = 0, ""
```

Comment:
this is confusing to me, is it:

```lua
local curr_max, curr_type, num = 0, "", nil
```

I'd much prefer to split this up and put it directly above the for loop (L382) but that's just personal preference. At least add the `, nil`